static/resources/harness.js

```diff
@@ -242,29 +243,34 @@
     }
   }
 
-  function runTest(data, i, callback) {
+  function runTest(data, i, oncomplete) {
     var test = data.tests[i];
 
+    function success(v) {
+      processTestResult(v, data, i, oncomplete);
+    }
+
+    function fail(e) {
+      var v;
+      if (e instanceof Error) {
+        v = e;
+      } else {
+        v = new Error(e);
+      }
+      processTestResult(v, data, i, oncomplete);
+    }
+
     try {
       var value = eval(test.code);
 
       if (typeof value === 'object' && value !== null && typeof value.then === 'function') {
-        value.then(
-            function(value) {
-              processTestResult(value, data, i, callback);
-            },
-            function(fail) {
-              processTestResult(new Error(fail), data, i, callback);
-            });
-        value['catch'](
-            function(err) {
-              processTestResult(err, data, i, callback);
-            });
-      } else {
-        processTestResult(value, data, i, callback);
+        value.then(success, fail);
+        value['catch'](fail);
```

Comment:
Is both this line and the previous needed? It seems like we should either have just `value.then(success, fail)` or `value.then(success)` + `value['catch'](fail)`. Is this due to one of those forms not being supported somewhere early on? If so a comment about it would be great.